### PR TITLE
Conversions from sbyte or short to nuint require explicit cast

### LINKED ISSUE #49596: sbyte/short to nuint conversions are implicit
**Version Used**: 3.8.0-5.20567.16 (53c5d7d3cf13d88978744a32a27c5f8350a8400a), Visual Studio 16.8.2

The spec at https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/native-integers.md says:

> | Operand | Target | Conversion | IL |
> |:---:|:---:|:---:|:---:|
> | `sbyte` | `nuint` | ExplicitNumeric | `conv.u` / `conv.ovf.u` |
> | `short` | `nuint` | ExplicitNumeric | `conv.u` / `conv.ovf.u` |

But the compiler allows these two conversions implicitly, which both change the value of negative input. If the casts are made explicit, IDE0004 is reported.

```csharp
static class NativeInts
{
    // These compile fine
    static nuint Implicit(sbyte x) => x;
    static nuint Implicit(short x) => x;

    // IDE0004 reported on both of these
    static nuint Explicit(sbyte x) => (nuint)x;
    static nuint Explicit(short x) => (nuint)x;
}
```

In a checked context, IDE0004 is still reported for explicit casts, and applying the code fixer affects the codegen.

```csharp
static class NativeInts
{
    static nuint CheckedExplicit(short x) => checked((nuint)x); // emits conv.ovf.u
    static nuint CheckedImplicit(short x) => checked(x);        // emits conv.i
}
```

I suspect that since C# 9 has shipped, this can no longer be fixed to match the current spec for compatibility reasons. That makes these the only built-in implicit conversions that can overflow. I think that at least IDE0004 should be updated to not suggest or offer to remove the "redundant" cast in a checked context.

### PR DESCRIPTION
**Pull request by cston**
Conversions from `sbyte` or `short` to `nuint`, and some binary operations between `sbyte` or `short` and `nuint`, require explicit casts.

Fixes #49596

### CODE REVIEW

**Comment by AlekseyTs** [comment]
>         F1 = int.MaxValue;

Should we be getting similar diagnostics here?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:5488 in 687096b. [](commit_id = 687096b963242d80e7a0e59412bba29a1f3800ed, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs [hunk 1]
```diff
@@ -6458,35 +6478,82 @@ .maxstack  1
...
+    static void Main()
+    {
+        F<ushort>(sbyte.MaxValue);
+        F<ushort>(short.MaxValue);
+        F<uint>(sbyte.MaxValue);
+        F<uint>(short.MaxValue);
+        F<uint>(int.MaxValue);
+        F<nuint>(sbyte.MaxValue);
+        F<nuint>(short.MaxValue);
+        F<nuint>(int.MaxValue);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:6519]: >int.MaxValue [](start = 17, length = 12)

Consider also testing ```long```. In other similar tests too.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Comment by cston** [comment]
> Should we be getting similar diagnostics here?

I don't think converting constants `sbyte.MaxValue` or `short.MaxValue` to `ushort` or `uint` or `nuint` should generate diagnostics. Currently converting `int.MaxValue` to `uint` or `nuint` is allowed which seems correct to me.

**Comment by AlekseyTs** [comment]
>  don't think converting constants `sbyte.MaxValue` or `short.MaxValue` to `ushort` or `uint` or `nuint` should generate diagnostics. Currently converting `int.MaxValue` to `uint` or `nuint` is allowed which seems correct to me.

It would be good to support this with a quote/link from the spec.

**Comment by cston** [comment]
> It would be good to support this with a quote/link from the spec.

From [implicit-constant-expression-conversions](https://github.com/dotnet/csharplang/blob/master/spec/conversions.md#implicit-constant-expression-conversions):

> An implicit constant expression conversion permits the following conversions:
> 
> A constant_expression (Constant expressions) of type int can be converted to type sbyte, byte, short, ushort, uint, or ulong, provided the value of the constant_expression is within the range of the destination type.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

**Review by 333fred — APPROVED** [review]
LGTM (commit 3)

**Pull request merged**